patches/server/0830-Saveable-Scoreboards.patch

```diff
@@ -0,0 +1,130 @@
++
++    private final Path file;
++    private final ScoreboardSaveData saveData;
++
++    public PaperSaveableScoreboard(ServerScoreboard serverScoreboard, Path file) {
++        super(serverScoreboard);
++        this.saveData = this.getHandle().createData();
++        if (Files.exists(file)) {
++            CompoundTag tag = null;
++            try {
++                tag = serverScoreboard.server.overworld().getDataStorage().readTagFromDisk(file.toFile(), SharedConstants.getCurrentVersion().getWorldVersion());
++            } catch (Exception e) {
++                throw new IllegalArgumentException("Could not load scoreboard from " + file, e);
++            }
++            this.saveData.load(tag.getCompound("data"));
++        }
++        this.file = file;
++    }
++
++    @Override
++    public synchronized void save() {
++        this.saveData.save(this.file.toFile());
++    }
++
++    @Override
++    public @NotNull Path getFile() {
++        return this.file;
++    }
++
++    @Override
++    public ServerScoreboard getHandle() {
++        return (ServerScoreboard) super.getHandle();
++    }
++
++}
+diff --git a/src/main/java/net/minecraft/world/level/storage/DimensionDataStorage.java b/src/main/java/net/minecraft/world/level/storage/DimensionDataStorage.java
+index e84b99ca10621cb1021d7ec17ba21df24e8c7474..46caeb6be996b481d8896444b7f698b212bd680b 100644
+--- a/src/main/java/net/minecraft/world/level/storage/DimensionDataStorage.java
++++ b/src/main/java/net/minecraft/world/level/storage/DimensionDataStorage.java
+@@ -78,6 +78,11 @@ public class DimensionDataStorage {
+ 
+     public CompoundTag readTagFromDisk(String id, int dataVersion) throws IOException {
+         File file = this.getDataFile(id);
++        // Paper start
++        return readTagFromDisk(file, dataVersion);
++    }
++    public CompoundTag readTagFromDisk(File file, int dataVersion) throws IOException {
++        // Paper end
+         FileInputStream fileInputStream = new FileInputStream(file);
+ 
+         CompoundTag var8;
+diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
+index 1a3b1eb7b70b9a668aa33ea943c13890eaa23a05..afea73e764b9f12997a8ceabc7933bbcd2a7a212 100644
+--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
++++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
+@@ -25,7 +25,7 @@ import org.bukkit.scoreboard.ScoreboardManager;
+ public final class CraftScoreboardManager implements ScoreboardManager {
+     private final CraftScoreboard mainScoreboard;
+     private final MinecraftServer server;
+-    private final Collection<CraftScoreboard> scoreboards = new WeakCollection<CraftScoreboard>();
++    private final Collection<CraftScoreboard> scoreboards = java.util.Collections.synchronizedCollection(new WeakCollection<CraftScoreboard>()); // Paper
```

Comment:
if we wanna use a synchronised collection you'd need to wrap the getScores method iteration in a sync block over this